File: NHF/sudoku.pl
% :- pred megoldase(sspec::in, ssol::in).
% megoldase(+SSpec,+SSol) : sikeres, ha az SSol érték-mátrix megoldása az SSpec Sudoku-feladványnak.
megoldase(s(K,Mx),Sol) :-
	valid(K,Sol),
	fitting(K,Mx,Sol).

% fitting(K,Mx,Sol) : sikeres ha Sol nem mond ellent a mezoinfoknak
fitting(K,Mx,Sol) :-
	N is K*K,
	myforall(
		(bet(1,N,Col),bet(1,N,Row)), 
		(mx_item(Mx,Row,Col,Infos),	infosHold(Infos,Sol,Row,Col))
	).

% infosHold(L,Sol,Row,Col) : sikres, ha Sol Row. soranak Col. eleme kielegiti L infolista minden elemet.
infosHold([H|T],Sol,Row,Col) :-
	infoHolds(H,Sol,Row,Col),
	infosHold(T,Sol,Row,Col).

infosHold([],_,_,_).

% infosHold(I,Sol,Row,Col) : sikres, ha Sol Row. soranak Col. eleme kielegiti I infot.
infoHolds(v(V),Sol,Row,Col) :-
	mx_item(Sol,Row,Col,V).
infoHolds(e,Sol,Row,Col) :-
	mx_item(Sol,Row,Col,V),
	0 is V mod 2.
infoHolds(o,Sol,Row,Col) :-
	mx_item(Sol,Row,Col,V),
	1 is V mod 2.
infoHolds(w,Sol,Row,Col) :-
	mx_item(Sol,Row,Col,V1),
	Col1 is Col - 1,
	mx_item(Sol,Row,Col1,V2),
	Sum is V1 + V2,
	1 is Sum mod 2.
infoHolds(s,Sol,Row,Col) :-
	mx_item(Sol,Row,Col,V1),
	Row1 is Row + 1,
	mx_item(Sol,Row1,Col,V2),
	Sum is V1 + V2,
	1 is Sum mod 2.

% valid(K,Sol) sikeres, ha Sol egy valid K parameteru sudoku kitoltes
valid(K,Sol) :- 
	valid_ll(K,Sol),
	N is K * K,
	feldarabolasa(Sol,N-1,ColL),
	valid_ll(K,ColL),
	feldarabolasa(Sol,K-K,SqL),
	valid_ll(K,SqL).

% valid_ll(K,LL) sikeres, ha LL minden eleme 1 tol K*K ig tartalmazza a szamok
% egy permutaciojat
valid_ll(K,[H|T]) :-
	valid_ll(K,T),
	N is K * K,	
	valid_list(N,H).
valid_ll(_,[]).
	

% valid_list(K,L) sikeres, ha L minden elemei a 1 tol K*K ig a szamok
% egy permutacioja
valid_list(N,L) :-
	length(L,N),
	myforall(bet(1,N,X), member(X,L)).

% myforall(A,B) sikeres, ha minden A ra B is igaz
myforall(A, B) :-
    \+ (call(A), \+ call(B)).	
	
%------------------------------   KHF2 ------------------------------

%mx_item(Mx,R,C,Val)
%Val Mx matrix R sorának C oszlopában talalhato eleme
mx_item(Mx,R,C,Val) :-
	list_nth(Mx,R,TMP),
	list_nth(TMP,C,Val).

%list_nth(L,N,Val)
%Val L lista N. eleme (1tol kezdve a szamozast)
list_nth([H|_],1,H).
list_nth([_|T],N,Val):-
	N > 1,
	N1 is N-1,
	list_nth(T,N1,Val).

%bet(From,To,N)
%N From es To kozotti szam
bet(From,To,N) :-
	From =< To, N = From.
bet(From,To,N) :-
	From < To,
	From1 is From + 1,
	bet(From1,To,N).

%------------------------------  /KHF2 ------------------------------
%------------------------------   KHF1 ------------------------------

% :- type matrix == list(row).
% :- type row == list(any).
% :- type parameter ---> subRows-subCols.
% :- type subRows == int.
% :- type subCols == int.
% :- pred feldarabolasa(+matrix, +parameter, ?list(list(any))).
% feldarabolasa(Mx, P, LL): Az LL lista az Mx mátrix P paraméterű feldarabolása.
feldarabolasa(Mx, R-C, LL) :- 
	feldarabolasa(Mx,R,C,LLt),
	flattenNextLvl(LLt,LL).


% feldarabolasa(Mx, R,C, LL): Az LL lista az Mx mátrix R-C paraméterű mátrixokra darabolása
feldarabolasa(Mx, R, C, LL):- 
	countR(Mx, R, Cr, 0),
	countC(Mx, C, Cc, 0),
	split(Mx, R, C, Cr, Cc, LL).


% split(Mx, R,C, Cr, Cc, LL): Az LL lista az Mx mátrix R-C paraméterű mátrixokra darabolása,
% sorok mentén legfeljebb Cr, oszolopok mentén legfeljebb Cc részre vágva
split([Hmx|Tmx], R,  C, Cr, Cc, LL) :-
	split([Hmx|Tmx], R, C, Cr, Cc, 0, 0, LL).

% split(Mx, R,C, Cr, Cc, Cntr, Cntc LL): Az LL lista az Mx mátrix R-C paraméterű mátrixokra darabolása,
% sorok mentén legfeljebb Cr, oszolopok mentén legfeljebb Cc részre vágva
% a sorok szerint Cntr, az oszlopok szerint Cntc-dik részmátrixtól kezdve
split(_, _, _, Cr, _, Cr, 0, []).
split([H|T],R,C,Cr,Cc,Cntr,Cc,L) :-
	Cntr1 is Cntr+1,
	split([H|T],R,C,Cr,Cc,Cntr1,0,L).
split([Hmx|Tmx], R,  C, Cr, Cc, Cntr, Cntc, [H1|T1]) :- 
	Cntr < Cr,
	Cntc < Cc,
	Cntc1 is Cntc + 1,
	StartC is Cntc * C + 1,
	StartR is Cntr * R + 1,
	submatrix([Hmx|Tmx], StartC, C, StartR, R, H1),
	split([Hmx|Tmx], R, C, Cr, Cc, Cntr, Cntc1, T1).

% submatrix(Mx, StartC, Width, StartR, Height, Smx):
%% Smx mátrix az Mx mátrix olyan sor és oszlopfolytonos részmátrixa,
%% ami a StartC oszloptól Width oszlop szélesen,
%% és a StartR sortól Height sor hosszan tart 
submatrix(Mx, StartC, Width, StartR, Height, Smx) :-
	sublist(Mx, StartR, Height, Lt),
	cutAll(Lt, StartC, Width, Smx).

% cutAll(Mx, StartR, Height, M) M mátrix Mx olyan részmátrixa, mely annak
% Height sorát tartalmazza sorfolytonosan a StartR. sortól kezdve	
cutAll([], _, _, []).
cutAll([H|T], StartR, Height, [HSmx|TSmx]) :-
	isList(H),
	sublist(H,StartR,Height, HSmx),
	cutAll(T, StartR, Height, TSmx).

% sublist(L, S, N, L1): Az L1 lista az L lista N hosszúságú részlistája az S. elemmel kezdve
sublist(L, S, N, L1) :-
	S1 is S -1,
	drop(L,S1,Ltemp),
	take(Ltemp,N,L1).

% take(L0, N, L): Az L lista az L0 lista N hosszú prefixuma
take([], _, []).
take([_|_], 0, []).
take([H|T], N, [H | Taken]) :-
	N > 0,
	N1 is N-1,
	take(T, N1, Taken).

% drop(L0, N, L): Az L0 lista olyan szuffixuma L, amely az L0 elsõ N elemét
%% nem tartalmazza
drop([], _, []).
drop([H|T], 0, [H|T]).
drop([_|T], N, Dropped) :-
	N > 0,
	N1 is N-1,
	drop(T,N1,Dropped).

% countC(Mx, C, Cc, Acc): Cc = Acc + ahány részre esik Mx mátrix C oszloponként vágva
countC([H|_], C, Cc, 0) :- 
	isList(H),
    countR(H,C,Cc,0).

%countR(Mx, R, Cr, Acc): Cr = Acc + ahány részre esik Mx mátrix R soronként vágva
countR([], _, 0, _).
countR([H|T], R, Cr, 0) :- 
	countR([H|T],R,UjCr,R),
	Cr is UjCr + 1.
countR([_|T], R, Cr, Rem) :-
	Rem > 0,
	UjRem is Rem -1,
	countR(T, R, Cr, UjRem).

% flattenNextLvl(L,Fl): Fl az L lista elemeit taralmazza, mindet kilapítva
flattenNextLvl([],[]).
flattenNextLvl([H0|T0],[H1|T1]):-
	flatten(H0,H1),
	flattenNextLvl(T0,T1).
	

% flatten(L,Fl) : Fl L lista kilapítva
flatten(L, Fl):-
  flatten(L, Fl, []).

% flatten(L, Fl, Acc) Fl az L lista kilapítva és Acc elé fűzve
flatten([], Fl, Fl).
flatten(H, [H|Fl], Fl):-
  \+ isList(H).
flatten([H|T], Fl, Acc):-
  flatten(H, Fl, L1),
  flatten(T, L1, Acc).

% isList(L) L egy tetszőleges lista (annak eldöntésére, hogy L lista-e)
isList([]).
isList([_|_]).

%------------------------------  /KHF1 ------------------------------

File: NHF/sudoku.erl
-module(sudoku).
-author('czipobence@gmail.com').
-vsn('2016-11-03').
%-export([sudoku/1]).
-compile(export_all).

%% @type sspec() = {size(), board()}.
%% @type size()  = integer().
%% @type field() = [info()].
%% @type info()  = e | o | s | w | integer().
%% @type board() = [[field()]].

%% @type ssol() = [[integer()]].

%% @spec sudoku:sudoku(SSpec::sspec()) -> SSols::[ssol()].
%% @doc  SSols az SSpec feladványt kielégítő megoldások listája.
sudoku({K,Bo}) -> 
	try solve(K,Bo) of
		Val -> Val
	catch 
		throw:no_solution -> []
	end.

solve(K,Bo) ->
	BoPP = preprocess(Bo),
	Vals = allowed(K,BoPP),
	process_singles(K,Vals,BoPP).
	
process_singles(K,Vals,BoPP) ->
	try process_one_single(Vals, BoPP,1) of 
		{R,C,V,I} -> 
			Result = setValue(K,Vals,R,C,V,I),
			process_singles(K,Result,BoPP)
	catch
		throw:no_more_single -> Vals
	end.

process_one_single([],_,_) -> throw(no_more_single);
process_one_single([HV|TV],[HB|TB],R) -> 
	try process_one_single_row(HV,HB,R,1) of
		Result -> Result
	catch
		throw:no_more_single -> process_one_single(TV,TB,R+1)
	end.
	
process_one_single_row([],_,_,_) -> throw(no_more_single);
process_one_single_row([[H]|_], [HB|_],R,C) -> {R,C,H,HB};
process_one_single_row([_|TV], [_|TB],R,C) -> 
	process_one_single_row(TV,TB,R,C+1).

setValue(K,Vals,R,C,V,I) ->
	lists:reverse(setValue(K,Vals,R,C,V,I,1,[])).
	
setValue(_,[],_,_,_,_,_,Acc) -> Acc;
setValue(K,[H|T],R,C,V,I,Rc,Acc) ->
	setValue(K,T,R,C,V,I,Rc+1, [lists:reverse(setValue_row(K,H,R,C,V,I,Rc,1,[]))|Acc]).
	
setValue_row(_,[],_,_,_,_,_,_,Acc) -> Acc;
setValue_row(K,[_|T],R,C,V,I,R,C,Acc) -> 
	setValue_row(K,T,R,C,V,I,R,C+1,[V|Acc]);
setValue_row(K,[H|T],R,C,V,I,Rc,C,Acc) when Rc == R-1 ->
	GuardS = lists:member(n,I),
	H1 = if 
		GuardS -> 
			if 
				V rem 2 == 0	-> filter_odds(H);
				true 			-> filter_evens(H)
			end;
		true -> H
	end,
	setValue_row(K,T,R,C,V,I,Rc,C+1,[delete_guess(V,H1)|Acc]);
setValue_row(K,[H|T],R,C,V,I,Rc,C,Acc) when Rc == R+1 ->
	GuardS = lists:member(s,I),
	H1 = if 
		GuardS -> 
			if 
				V rem 2 == 0	-> filter_odds(H);
				true 			-> filter_evens(H)
			end;
		true -> H
	end,
	setValue_row(K,T,R,C,V,I,Rc,C+1,[delete_guess(V,H1)|Acc]);
setValue_row(K,[H|T],R,C,V,I,R,Cc,Acc) when Cc == C+1 ->
	GuardS = lists:member(a,I),
	H1 = if 
		GuardS -> 
			if 
				V rem 2 == 0	-> filter_odds(H);
				true 			-> filter_evens(H)
			end;
		true -> H
	end,
	setValue_row(K,T,R,C,V,I,R,Cc+1,[delete_guess(V,H1)|Acc]);
setValue_row(K,[H|T],R,C,V,I,R,Cc,Acc) when Cc == C-1 ->
	GuardS = lists:member(w,I),
	H1 = if 
		GuardS -> 
			if 
				V rem 2 == 0	-> filter_odds(H);
				true 			-> filter_evens(H)
			end;
		true -> H
	end,
	setValue_row(K,T,R,C,V,I,R,Cc+1,[delete_guess(V,H1)|Acc]);

setValue_row(K,[H|T],R,C,V,I,R,Cc,Acc) -> 
	setValue_row(K,T,R,C,V,I,R,Cc+1,[delete_guess(V,H)|Acc]);
setValue_row(K,[H|T],R,C,V,I,Rc,C,Acc) -> 
	setValue_row(K,T,R,C,V,I,Rc,C+1,[delete_guess(V,H)|Acc]);
setValue_row(K,[H|T],R,C,V,I,Rc,Cc,Acc) when 
		(((R - (R-1) rem K) == (Rc - (Rc - 1) rem K)) andalso 
		((C - (C-1) rem K) == (Cc -(Cc-1) rem K))) -> 
	setValue_row(K,T,R,C,V,I,Rc,Cc+1,[delete_guess(V,H)|Acc]);
setValue_row(K,[H|T],R,C,V,I,Rc,Cc,Acc) -> 
	setValue_row(K,T,R,C,V,I,Rc,Cc+1,[H|Acc]).

delete_guess(_, []) -> throw(no_solution);
delete_guess(V, [V]) -> throw(no_solution);
delete_guess(V, [H|T]) -> lists:delete(V,[H|T]);
delete_guess(V, V) -> throw(no_solution);
delete_guess(_,N) -> N.

%% @spec sudoku:preprocess(K::integer(), M::board()) -> M1::board()
%% M1 egy tabla, ami a feldolgozott infokat tartaknazza, kiegeszitve 
%% azokat north (n) es east (a) informaciokkal
preprocess(Bo) -> lists:reverse(preprocess([],Bo,[])).

preprocess(_,[],Acc) -> Acc;
preprocess(PrevR,[CurrR|T], Acc) ->
	preprocess(CurrR,T,[lists:reverse(preprocess_row(PrevR,CurrR,[])) | Acc]).
	
preprocess_row(_,[],Acc) -> Acc;
preprocess_row([], [CurrI], Acc) -> [CurrI | Acc];
preprocess_row([], [CurrI,NextI|CurrT], Acc) -> 
	GuardW = lists:member(w,NextI),
	if 	GuardW 	-> preprocess_row([], [NextI | CurrT], [ [a | CurrI] | Acc ]);
		true 	-> preprocess_row([], [NextI | CurrT], [CurrI | Acc])
	end;
preprocess_row([PreI], [CurrI], Acc) -> 
	Guard = lists:member(s,PreI),
	if 	Guard -> [[n | CurrI]| Acc];
		true -> [CurrI | Acc]
	end;
preprocess_row([PreI|PreT], [CurrI,NextI|CurrT], Acc) ->
	GuardS = lists:member(s,PreI),
	GuardW = lists:member(w,NextI),
	CurrI1 = if	GuardS -> [n | CurrI];
		true -> CurrI
	end,
	if GuardW -> preprocess_row(PreT, [NextI|CurrT], [ [a | CurrI1] | Acc ]);
		true -> preprocess_row(PreT, [NextI|CurrT], [CurrI1 | Acc])
	end.

%% @spec sudoku:allowed(K::integer(), M::board()) -> Vals::[[[integer()]]]
%%	 Vals a Bo infoi altal megengedett mezoertekek matrixa
allowed(K,Bo) -> [[ parse_ertek(ertekek(K,Bo,R,C))
				  || C <- lists:seq(1,K*K)]
				  || R <- lists:seq(1,K*K) ].

parse_ertek([]) -> throw(no_solution);
parse_ertek(L) -> L.

%% @spec sudoku:ertekek(K::integer(), M::board(), R::integer(), C::integer()) -> Vals::[integer()]
%%   Vals a {K,M} specifikációval megadott Sudoku-feladvány R-C
%%   koordinátájú mezőjében megengedett értékek listája.
ertekek(K, M, R,C) -> 
	Val = nth_matr(M,R,C),
	Mx = set_nth_matr(M,R,C,[]),
	AllNum = lists:seq(1,K*K),
	L = merge(flatten(get_row(Mx,R)), merge(flatten(get_col(Mx,C)), flatten(submatrix(Mx, top(R,K), top(C,K) ,K,K)))),
	Allowed = lists:foldr(fun(X,Acc)->intersection(X,Acc) end, AllNum, lists:map(fun(X) -> allowed_by(X,K) end,L)),
	Limited = lists:foldr(fun(X,Acc)->intersection(X,Acc) end, AllNum, lists:map(fun(X) -> limited_to(X,K) end,Val)),
	intersection(Allowed,Limited).

%% @type info()  = e | o | s | w | integer().
%% @spec sudoku:limited_to(V::field(), K::integer()) -> L::[integer()].
%% L lista azon szamok listaja, amelyek V info mellett egy K parameteru sudoku adott mezojeben lehetnek
limited_to(V,K) when is_number(V) -> 
	case V < K*K+1 of
	true -> [V];
	false -> []
	end;
limited_to(o,K) -> lists:reverse(secs_till(K*K,1,[]));
limited_to(e,K) -> lists:reverse(secs_till(K*K,2,[]));
limited_to(_,K) -> lists:seq(1,K*K).

%% @spec sudoku:allowed_by(V::field(), K::integer()) -> L::[integer()].
%% L lista azon szamok listaja, amelyek V info mellett egy K parameteru 
%% sudokuban az adott mezovel egy szegmensben (K*K negyzet, sor, oszlop)
%% engedelyezettek
allowed_by(V,K)when is_number(V) -> lists:reverse(miss(V,K*K));
allowed_by(_,K) -> lists:seq(1,K*K).


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Sudoku related
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% @spec sudoku:top(R::integer(), K::integer()) -> Top::integer().
%% Top a legnagyobb, R-nel kisebb K-val oszthato poz. eg. szam
top(R,K) -> top(R,K,1).
top(R,K,Acc) ->
	case Acc + K > R of
	true -> Acc;
	false -> top(R,K,Acc+K)
	end.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Matrixok
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% @spec sudoku:get_row(Mx::matrix(), Row::integer()) -> L::[any()].
%% Az L lista Mx matrix Row. oszlopa
get_row(Mx, Row) -> lists:nth(Row, Mx).

%% @spec sudoku:get_col(Mx::matrix(), Col::integer()) -> L::[any()].
%% Az L lista Mx matrix Col. oszlopanak elemei (oszlop helyett listava lapitva)
get_col(Mx, Col) -> [lists:nth(Col,L) || L <- Mx].

%% @type matrix() = [row()].
%% @type row() = [any()].
%% @spec sudoku:nth_matr(M::matrix(), Row::integer(),Col::integer()) -> T::any().
%% T M matrix Row. soraban es Col. oszlopaban talalhato eleme
nth_matr(M, Row, Col) -> lists:nth(Col, lists:nth(Row, M)).

%% @spec sudoku:set_nth_matr(M::matrix(), Row::integer(),Col::integer(),To::any()) -> Mx::matrix().
%% Mx olyan matrix, ami megegyezik M-el, kiveve Row soranak Col oszlopa, ahol az elem To
set_nth_matr(M, Row,Col,To) -> [
		[set_nth_proc_element(nth_matr(M,R,C), R, C, Row, Col, To) 
						|| C <- lists:seq(1,length(hd(M)))] 
				|| R <- lists:seq(1,length(M))].

%% @spec sudoku:set_nth_proc_element(E::any(), R::integer(),C::integer(), Row::integer(), Col::integer(),To::any()) -> E1::matrix().
%% E1 az E, kiveve ha Row = R es Col = C, mert akkor To
set_nth_proc_element(_,R,C,R,C,To) -> To;
set_nth_proc_element(E,_,_,_,_,_) -> E.

%% @spec sudoku:submatrix(Mx::matrix(), StartX::integer(), StartY::integer(),Width::subRows(), Height::subCols()) -> M::matrix()
%% M mátrix az Mx mátrix olyan sor és oszlopfolytonos részmátrixa,
%% ami a StartX oszloptól Width oszlop szélesen,
%% és a StartY sortól Height sor hosszan tart 									
submatrix(Mx, StartX, StartY, Width, Height) ->
	[lists:sublist(L, StartY, Height) || L <- lists:sublist(Mx, StartX, Width)].
	
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Listakezeles
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%	
filter_odds([]) -> [];
filter_odds([H|T]) ->
	if
		(H rem 2) == 0 -> filter_odds(T);
		true -> [H| filter_odds(T)]
	end;
filter_odds(N) ->
	if
		N rem 2 == 0 -> 
			throw(no_solution);
		true ->
			N
	end.

filter_evens([]) -> [];
filter_evens([H|T]) ->
	if 
		(H rem 2 == 1) -> filter_evens(T);
		true -> [H | filter_evens(T)]
	end;
filter_evens(N) ->
	if
		N rem 2 == 1 -> 
			throw(no_solution);
		true ->
			N
	end.

%% @spec sudoku:intersection(L1::[any()], L2::[any()]) -> L::[any()].
%% L L1 L2 szigoruan monoton novekvo listak metszete
intersection(L1,L2) -> lists:reverse(intersection(L1,L2,[])).

%% @spec sudoku:intersection(L1::[any()], L2::[any()], Coll::[any()]) -> L::[any()].
%% L L1 L2 szigoruan monoton novekvo listak metszete osszefuzve Coll-al
intersection([],_,Coll) -> Coll;
intersection(_, [], Coll) -> Coll;
intersection([H|T1], [H|T2], Coll) -> intersection(T1,T2,[H|Coll]);
intersection([H1|T1], [H2|T2], Coll) when H1 < H2 -> 
		intersection(T1, [H2|T2], Coll);
intersection([H1|T1], [_|T2], Coll) -> 
		intersection([H1|T1], T2, Coll).

%% @spec sudoku:secs_till(N::integer(), I::integer(), L::[integer()]) -> L1::[integer()].
%% L lista I tol kezdve N-ig minden masodik szam listaja, megforditva L ele fuzva
secs_till(N,I,L) ->
	case I < N+1 of
	true -> secs_till(N,I+2,[I|L]); 
	false -> L
	end.

%% @spec sudoku:miss(V::integer(), N::integer()) -> L::[integer()].
%% L lista N tol 1-ig a szamok, V-t kihagyva
miss(V,N) -> miss(V,N,1,[]).

%% @spec sudoku:miss(V::integer(), N::integer(), Curr:integer(), L:[integer()]) -> L1::[integer()].
%% L1 lista N-tol Curr-ig a szamok, V-t kihagyva, L ele fuzve
miss(N,N,N,L) -> L;
miss(_,N,N,L) -> [N|L];
miss(V,N,V,L) -> miss(V,N,V+1,L);
miss(V,N,I,L) -> miss(V,N,I+1, [I|L]).
	
%% @spec sudoku:merge(L0::[any()], L1::[any()]) -> L::[any()].
%% L lista L0 es L1 osszefuzve
merge([],L) -> L;
merge([H|T], L) -> merge(T, [H|L]).

%% @spec sudoku:flatten(Mx::matrix()) -> LL::[any()].
%% Az LL lista az Mx mátrix kilapítva
flatten(Mx) -> flatten(Mx, []).


%% @spec sudoku:flatten(Mx::matrix(), L::[term()]) -> L1::[term()].
%% Az L1 lista az Mx mátrix kilapítva és L lista mögéfûzve
flatten([],L) -> L;
flatten([H|T], L) when is_list(H) -> flatten(H, flatten(T,L));
flatten([H|T], L) -> [H | flatten(T,L)].
